### PR TITLE
docs/tests/enhancements 

### PR DESCRIPTION
**Pull request by threepointone**
## Summary

This PR adds comprehensive documentation for the Agents SDK and implements several feature improvements across state management, MCP, routing, scheduling, and the callable RPC system.

**Highlights:**

- 15+ new documentation files covering core concepts and features
- 10+ new test files with 100+ tests
- New features: synchronous `setState()`, `validateStateChange()` hook, `scheduleEvery()`, `basePath` routing, server-sent identity, callable improvements
- No breaking changes (backward compatibility maintained)

---

### Documentation Index

| File | Description |
|------|-------------|
| `docs/index.md` | Main table of contents organizing all documentation by category |

### Getting Started & Integration

| File | Description |
|------|-------------|
| `docs/getting-started.md` | Quick start guide covering installation, first agent, state basics, and deployment |
| `docs/adding-to-existing-project.md` | Guide for integrating agents into existing Workers, including React apps, Hono, static assets, and common patterns |

### Core Concepts

| File | Description |
|------|-------------|
| `docs/state.md` | State management covering `initialState`, `setState()`, `validateStateChange()`, `onStateUpdate()`, persistence, and client synchronization |
| `docs/routing.md` | URL routing patterns, `routeAgentRequest()`, custom paths with `basePath`, server-sent identity, and advanced routing scenarios |
| `docs/http-websockets.md` | HTTP/WebSocket lifecycle hooks (`onRequest`, `onConnect`, `onMessage`, `onClose`), connection management, per-connection state, hibernation, and common patterns |

### Features

| File | Description |
|------|-------------|
| `docs/callable-methods.md` | `@callable` decorator, RPC over WebSocket, streaming responses, TypeScript integration, error handling, and when to use DO RPC instead |
| `docs/mcp-client.md` | Connecting to MCP servers with `addMcpServer()`, OAuth flows, transport options, and tool usage |
| `docs/scheduling.md` | One-time, recurring (`scheduleEvery`), and cron-based scheduling |

---

### State Management

**Synchronous `setState()`** - `setState()` is now synchronous, returning `void` instead of `Promise<void>`. Existing code using `await this.setState(...)` continues to work.

```typescript
// Before (still works - awaiting void is harmless)
await this.setState({ count: 1 });

// After (preferred)
this.setState({ count: 1 });
```

**`validateStateChange()` hook** - New synchronous validation hook that runs before state is persisted and broadcast. Throw to reject updates:

```typescript
validateStateChange(nextState: State, source: Connection | "server") {
  if (nextState.count < 0) {
    throw new Error("Count cannot be negative");
  }
}
```

**`onStateUpdate()` is now non-gating** - Runs via `ctx.waitUntil()` after persistence and broadcast. Errors are routed to `onError()` but don't prevent state updates.

**Execution order:**
1. `validateStateChange(nextState, source)` - validation (sync, gating)
2. State persisted to SQLite
3. State broadcast to connected clients
4. `onStateUpdate(nextState, source)` - notifications (async via `ctx.waitUntil`, non-gating)

### Scheduling

**`scheduleEvery(intervalSeconds, callback, payload?)`** - Fixed-interval recurring tasks with overlap prevention and error resilience.

```typescript
await this.scheduleEvery(60, "cleanup");
await this.scheduleEvery(300, "syncData", { source: "api" });
```

- Validates interval doesn't exceed 30 days (DO alarm limit)
- Overlap prevention with hung callback detection (configurable via `hungScheduleTimeoutSeconds`)

### MCP Server API

**Options-based `addMcpServer()` overload** - Cleaner API avoiding positional `undefined` parameters.

```typescript
// New preferred format
await this.addMcpServer("server", url, {
  callbackHost: "https://my-worker.workers.dev",
  transport: { headers: { Authorization: "Bearer ..." } }
});
```

### Custom Routing

- **`basePath`** - Bypass default URL construction for custom routing
- **`path`** - Append sub-paths to routes
- **Server-sent identity** - Agents send `name` and `agent` type on connect
- **`onIdentity` / `onIdentityChange`** - Callbacks for identity events
- **`identified` / `ready`** - State for identity status
- **`static options = { sendIdentityOnConnect }`** - Server-side control

```typescript
const agent = useAgent({
  basePath: "user",
  onIdentity: (name, agentType) => console.log(`Connected to ${name}`)
});
```

### Callable System Improvements

- **Client-side RPC timeout** - `{ timeout: 5000 }`
- **`StreamingResponse.error(message)`** - Graceful stream error signaling
- **`getCallableMethods()`** - Introspection API for callable methods
- **Connection close handling** - Pending calls rejected on disconnect
- **WeakMap for metadata** - Prevents memory leaks
- **`crypto.randomUUID()`** - More robust RPC IDs
- **Streaming observability** - Events for streaming calls
- **Streaming error logging** - Server-side errors now logged for observability

```typescript
// New format (supports timeout)
await agent.call("method", [args], {
  timeout: 5000,
  stream: { onChunk, onDone, onError }
});

// Legacy format (still supported)
await agent.call("method", [args], { onChunk, onDone, onError });
```

### Email Utilities

- **`isAutoReplyEmail(headers)`** - Detect auto-reply emails using standard RFC headers (`Auto-Submitted`, `X-Auto-Response-Suppress`, `Precedence`)

---

## Tests Added

| File | Coverage |
|------|----------|
| `packages/agents/src/tests/state.test.ts` | State persistence, sync, validation, edge cases |
| `packages/agents/src/tests/basepath.test.ts` | Custom path routing |
| `packages/agents/src/tests/routing.test.ts` | General routing (22 tests) |
| `packages/agents/src/tests/callable.test.ts` | RPC, streaming, errors, introspection (17 tests) |
| `packages/agents/src/tests/client-timeout.test.ts` | Client-side RPC timeouts |
| `packages/agents/src/tests/message-handling.test.ts` | WebSocket message handling |
| `packages/agents/src/tests/email-headers.test.ts` | Email header utilities |
| `packages/agents/src/tests/mcp/add-mcp-server.test.ts` | MCP options API, legacy signature |

---

## Bug Fixes

- **React `useRef` type fix** - Fixed type error in `react.tsx`
- **Memory leak prevention** - WeakMap for callable metadata storage
- **Connection cleanup** - Pending RPC calls rejected on WebSocket close
- **JSON parse error handling** - Graceful fallback to `initialState` on corrupted state
- **Streaming error logging** - Server-side errors in streaming methods are now logged

---

### Key Areas for Review

1. **State Management** (`packages/agents/src/index.ts`)
   - Synchronous `setState()` refactor
   - `validateStateChange()` hook
   - `onStateUpdate()` via `ctx.waitUntil()`

2. **Routing & Identity** (`packages/agents/src/index.ts`, `client.ts`, `react.tsx`)
   - New `CF_AGENT_IDENTITY` WebSocket message type
   - Client state management for identity
   - `static options` pattern for server-side configuration

3. **Callable System** (`packages/agents/src/client.ts`)
   - Backward compatibility detection for streaming options
   - Timeout implementation with cleanup

4. **Schema Migration** (`scheduleEvery`)
   - Adds `intervalSeconds`, `running`, and `execution_started_at` columns to `cf_agents_schedules`
   - Auto-migrates existing agents

### Test Commands

```bash
# Run all tests
CI=true npm test

# Run specific test files
cd packages/agents && npm run test:workers -- callable.test.ts
cd packages/agents && npm run test:workers -- routing.test.ts
cd packages/agents && npm run test:workers -- state.test.ts
```

### Changesets

- `.changeset/sync-set-state.md` - Synchronous setState refactor
- `.changeset/callable-improvements.md` - Callable system enhancements
- `.changeset/identity-routing-enhancements.md` - Routing and identity features
- `.changeset/friendly-mcp-options.md` - MCP options API
- `.changeset/cool-intervals-run.md` - scheduleEvery method
- `.changeset/email-header-utilities.md` - Email header utilities
- `.changeset/fix-serializable-interface-types.md` - Serializable type fixes
- `.changeset/bug-fixes-three.md` - Bug fixes

### CODE REVIEW

File: .changeset/identity-routing-enhancements.md [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+useAgent({ agent: "MyAgent", name: "room", path: "settings" });
+```
+
+## Server-Side Identity Control
+
+Disable identity sending for security-sensitive instance names:
+
+```typescript
+class SecureAgent extends Agent {
+  static agentOptions = { sendIdentityOnConnect: false };
```
Comment (threepointone) [inline .changeset/identity-routing-enhancements.md]: hmm this is options, not agentOptions

File: .changeset/callable-improvements.md [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+  stream.end();
+}
+```
+
+### getCallableMethods() API
+
+New method on the Agent class to introspect all callable methods and their metadata:
+
+```typescript
+const methods = agent.getCallableMethods();
```
Comment (threepointone) [inline .changeset/callable-improvements.md:44]: @craigsdennis you wanted this

File: .changeset/friendly-mcp-options.md [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+await this.addMcpServer("server", url, undefined, undefined, {
+  transport: { headers: { Authorization: "Bearer ..." } }
+});
+```
+
+### After (preferred)
+
+```typescript
+// Clean options object
+await this.addMcpServer("server", url, {
```
Comment (threepointone) [inline .changeset/friendly-mcp-options.md:22]: @mattzcarey better, no?

File: .changeset/identity-routing-enhancements.md [hunk 4]
```diff
@@ -0,0 +1,106 @@
+---
+"agents": patch
+---
+
+Add custom URL routing with `basePath` and server-sent identity
+
+## Custom URL Routing with `basePath`
+
+New `basePath` option bypasses default `/agents/{agent}/{name}` URL construction, enabling custom routing patterns:
```
Comment (threepointone) [inline .changeset/identity-routing-enhancements.md:9]: @whoiskatrin ramp wanted this

File: docs/getting-started.md [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+  count: number;
+};
+
+// Create the agent
+export class Counter extends Agent<Env, CounterState> {
+  // Initial state for new instances
+  initialState: CounterState = { count: 0 };
+
+  // Methods can be called from the client
+  increment() {
```
Comment (threepointone) [inline docs/getting-started.md:54]: these methods need `@callable `

File: packages/agents/src/index.ts [hunk 6]
```diff
@@ -397,13 +462,40 @@ export class Agent<
...
 
+  /**
+   * Agent-specific configuration options.
+   * Override in subclasses - only specify what you want to change.
+   * @example
+   * class SecureAgent extends Agent {
+   *   static agentOptions = { sendIdentityOnConnect: false };
+   * }
+   */
+  static agentOptions: Partial<Omit<AgentStaticOptions, "hibernate">> = {};
```
Comment (threepointone) [inline packages/agents/src/index.ts]: sus

File: packages/agents/src/index.ts [hunk 7]
```diff
@@ -397,13 +462,40 @@ export class Agent<
...
+
+  /**
+   * Resolved options (merges defaults with subclass overrides)
+   */
+  private get _resolvedOptions(): AgentStaticOptions {
```
Comment (threepointone) [inline packages/agents/src/index.ts]: downstream of sus

File: packages/agents/src/index.ts [hunk 8]
```diff
@@ -2765,13 +3111,21 @@ export class StreamingResponse {
...
   /**
    * Send a chunk of data to the client
    * @param chunk The data to send
+   * @returns false if stream is already closed (no-op), true if sent
    */
-  send(chunk: unknown) {
+  send(chunk: unknown): boolean {
     if (this._closed) {
-      throw new Error("StreamingResponse is already closed");
+      return false;
```
Comment (threepointone) [inline packages/agents/src/index.ts:3575]: should we log a warning here

File: docs/getting-started.md [hunk 9]
```diff
@@ -0,0 +1,266 @@
...
+
+Key points:
+
+- **`useAgent`** connects to your agent via WebSocket
+- **`onStateUpdate`** fires whenever the agent's state changes
+- **`agent.stub.methodName()`** calls methods on your agent
+
+---
+
+## What Just Happened?
```
Comment (irvinebroque) [inline docs/getting-started.md:153]: Love this as next thing right after initial setup, something we should steal for all our docs — get you going straight away and then do something interactive with what you've made — then _immediately_ explain step by step the mechanics of how it works (without writing paragraphs, and with a diagram)

File: .changeset/friendly-mcp-options.md (hunk 3 above)
Comment (whoiskatrin) [reply .changeset/friendly-mcp-options.md:22]: i am actually a fan of this, i like it, it's clean

File: docs/index.md [hunk 10]
```diff
@@ -1,23 +1,94 @@
...
+
+- [Client SDK](./client-sdk.md) - Connecting from React (`useAgent`) and vanilla JS (`AgentClient`), state sync, and RPC calls
+
+## Communication Channels
+
+- [Email Routing](./email.md) - Receiving and responding to emails
+- [Webhooks](./webhooks.md) - Receiving and sending webhook events
+- TODO: [SMS](./sms.md) - Text message integration (Twilio, etc.)
+- TODO: [Voice & Audio](./voice-audio.md) - Voice calls, audio streaming, and speech-to-text
+- TODO: [Messengers](./messengers.md) - Slack, Discord, Telegram, and other chat platforms
```
Comment (whoiskatrin) [inline docs/index.md:27]: maybe we should link to discord and slack agents in awesome-agents?

File: docs/index.md [hunk 11]
```diff
@@ -1,23 +1,94 @@
...
+
+## Migration Guides
+
+- [Migration to AI SDK v5](./migration-to-ai-sdk-v5.md)
+- [Migration to AI SDK v6](./migration-to-ai-sdk-v6.md)
+
+## Reference
+
+- TODO: [API Reference](./api-reference.md) - Complete API documentation
+- TODO: [FAQ / How is this different from Durable Objects?](./faq.md)
```
Comment (whoiskatrin) [inline docs/index.md:87]: you owe this .md to the world

File: docs/mcp-client.md [hunk 12]
```diff
@@ -0,0 +1,502 @@
...
+
+```typescript
+import { Agent } from "agents";
+
+export class MyAgent extends Agent {
+  async onRequest(request: Request) {
+    // Add an MCP server
+    const result = await this.addMcpServer(
+      "github",
+      "https://mcp.github.com/sse"
```
Comment (whoiskatrin) [inline docs/mcp-client.md]: we should use /mcp?

File: docs/mcp-client.md [hunk 13]
```diff
@@ -0,0 +1,502 @@
...
+
+```typescript
+const result = await this.addMcpServer(name, url, options?);
+```
+
+### Basic Usage
+
+```typescript
+// Simple connection
+await this.addMcpServer("notion", "https://mcp.notion.so/sse");
```
Comment (whoiskatrin) [inline docs/mcp-client.md]: same /mcp

File: docs/mcp-client.md [hunk 14]
```diff
@@ -0,0 +1,502 @@
...
+
+// With explicit callback host
+await this.addMcpServer("github", "https://mcp.github.com/sse", {
```
Comment (whoiskatrin) [inline docs/mcp-client.md]: /mcp

File: docs/mcp-client.md [hunk 15]
```diff
@@ -0,0 +1,502 @@
...
+
+### Transport Options
+
+MCP supports multiple transport types:
+
+```typescript
+await this.addMcpServer("server", "https://mcp.example.com", {
+  transport: {
+    // Transport type: "sse", "streamable-http", or "auto" (default)
+    type: "sse"
```
Comment (whoiskatrin) [inline docs/mcp-client.md]: streamable-http

File: docs/mcp-client.md [hunk 16]
```diff
@@ -0,0 +1,502 @@
...
+    MCPServer->>Agent: Callback with code
+    Agent->>MCPServer: Exchange for token
+    Agent-->>Client: onMcpUpdate (ready)
+```
+
+### Handling OAuth in Your Agent
+
+```typescript
+async onRequest(request: Request) {
+  const result = await this.addMcpServer("github", "https://mcp.github.com/sse");
```
Comment (whoiskatrin) [inline docs/mcp-client.md]: /mcp

File: docs/mcp-client.md (hunk 12 above)
Comment (threepointone) [reply docs/mcp-client.md]: fixed all the sse references, great catch, thank you!

File: .changeset/email-header-utilities.md [hunk 17]
```diff
@@ -0,0 +1,38 @@
...
+- `getEmailHeader(headers, name)` - Gets a specific header value (case-insensitive)
+- `hasEmailHeader(headers, name)` - Checks if a header exists (case-insensitive)
+- `hasAnyEmailHeader(headers, names)` - Checks if any of the specified headers exist
+- `getAllEmailHeaders(headers, name)` - Gets all values for headers that appear multiple times (e.g., `Received`)
+- `isAutoReplyEmail(headers, subject?)` - Detects auto-reply emails based on common headers and subject patterns
+
+Example usage:
+
+```typescript
+import { parseEmailHeaders, isAutoReplyEmail } from "agents";
```
Comment (threepointone) [inline .changeset/email-header-utilities.md]:
```suggestion
import { parseEmailHeaders, isAutoReplyEmail } from "agents/email";
```

File: packages/agents/src/index.ts [hunk 18]
```diff
@@ -759,7 +913,7 @@ export class Agent<
     }
   }
 
-  private _setStateInternal(
+  private async _setStateInternal(
```
Comment (threepointone) [inline packages/agents/src/index.ts]: oh hmm this isn't ideal

Comment (threepointone) [reply packages/agents/src/index.ts]: fixed

File: docs/index.md (hunk 10 above)
Comment (threepointone) [reply docs/index.md:27]: yes cap, good shout

File: docs/index.md (hunk 11 above)
Comment (threepointone) [reply docs/index.md:87]: don't I know it
